tempesta_fw/sock.c

```diff
@@ -432,6 +432,102 @@ ss_close_sync(struct sock *sk)
 }
 EXPORT_SYMBOL(ss_close_sync);
 
+/*
+ * Process a single SKB.
+ */
+static int
+ss_tcp_process_skb(struct sock *sk, struct sk_buff *skb, int *processed)
+{
+	int r = 0;
+	bool tcp_fin;
+	unsigned int offset, count;
+	void *conn;
+	struct sk_buff *frag_list;
+	struct tcp_sock *tp = tcp_sk(sk);
+
+	/* Calculate the offset into the SKB. */
+	offset = tp->copied_seq - TCP_SKB_CB(skb)->seq;
+	if (TCP_SKB_CB(skb)->tcp_flags & TCPHDR_SYN)
+		offset--;
+
+	/* SKB may be freed in processing. Save the flag. */
+	tcp_fin = TCP_SKB_CB(skb)->tcp_flags & TCPHDR_FIN;
+
+	/*
+	 * When GRO is used, multiple SKBs may be merged into
+	 * one big SKB. These SKBs are linked in via frag_list.
+	 * Interpret the big SKB as a set of separate smaller
+	 * SKBs for processing. Make the top SKB first in the
+	 * frag_list.
+	 */
+	frag_list = skb_shinfo(skb)->frag_list;
+	if (frag_list) {
+		struct sk_buff *f_skb;
+		skb_walk_frags(skb, f_skb) {
+			f_skb->nohdr = 0;
+			ss_skb_adjust_data_len(skb, -f_skb->len);
+		}
+		skb_shinfo(skb)->frag_list = NULL;
+	}
+	skb->next = frag_list;
+	frag_list = skb;
+
+	while (frag_list) {
+		skb = frag_list;
+		frag_list = frag_list->next;
```

Comment:
```skb``` is never NULL, you assign ```skb``` to ```frag_list```, which is just a loop iterator, check it for NULL and reassign ```frag_list``` to ```skb``` - looks tricky and ugly. Why not to use plain for() loop?